bundles/net.tourbook.device.garmin.fit/src/net/tourbook/device/garmin/fit/listeners/MesgListener_Session.java

```diff
@@ -105,16 +107,36 @@ public void onMesg(final SessionMesg mesg) {
       }
 
       // -----------------------POWER -----------------------
-
-      final Integer avgPower = mesg.getAvgPower();
+      Integer avgPower = mesg.getAvgPower();
       if (avgPower != null) {
          tourData.setPower_Avg(avgPower);
       }
-      final Integer maxPower = mesg.getMaxPower();
+
+      Integer maxPower = mesg.getMaxPower();
       if (maxPower != null) {
          tourData.setPower_Max(maxPower);
       }
 
+      // Looking if the power was retrieved from the developer fields
+      if (mesg.getMaxPower() == null && mesg.getAvgPower() == null) {
+         final ArrayList<Float> powerDataList = new ArrayList<>();
+         for (final TimeData timeData : fitData.getAllTimeData()) {
+            powerDataList.add(timeData.power);
+
+            if (tourData.getPower_DataSource() == null) {
+               tourData.setPower_DataSource(timeData.powerDataSource);
+            }
+         }
+
+         avgPower = (int) powerDataList.stream().mapToDouble(Float::doubleValue).average().getAsDouble();
```

Comment:
@wolfgang-ch 
I didn't use what you recommended "tourData.computeAvg_FromValues" as at this stage, the time series not having been calculated, the computeAvg_FromValues would return 0 [here ](https://github.com/wolfgang-ch/mytourbook/blob/5bd78f2e7c88154a49bf6d091b3b6c8ff1cc803f/bundles/net.tourbook/src/net/tourbook/data/TourData.java#L2841)

Reply:
this average() method do propably not take into account when you have paused and also the time between 2 time slices, which could be > 10 sec or 1 sec

Reply:
Correct, but again, I can't use this tourData.computeAvg_FromValues because the time series is null

Reply:
I am looking if I can use the function you advised as you are right, we would want to take into account when the user has paused and also the time between 2 time slices, which could be > 10 sec or 1 sec
If you have a better idea, let me know

Reply:
Maybe I should set a boolean such as computeAveragePower in the clas FitData.java
and if this boolean is true, I compute the average power [here ](https://github.com/wolfgang-ch/mytourbook/blob/5bd78f2e7c88154a49bf6d091b3b6c8ff1cc803f/bundles/net.tourbook.device.garmin.fit/src/net/tourbook/device/garmin/fit/FitData.java#L176) ??

Reply:
This is the location where I also have computed the power, after tourdata is filled up

As you can see in the other boolean fields in FitData, they have a prefix _is... which I use to start boolean fields, for me it is easier to read it and I also try to group them together when it makes sense.

Reply:
Ok, I have renamed it to _isComputeAveragePower